File: NodeData.py
# NodeData.py

from dataclasses import dataclass, asdict, field
from typing import List

@dataclass
class Serializable:
    def to_dict(self):
        return asdict(self)
    
    @classmethod
    def from_dict(cls, data):
        return cls(**data)

@dataclass
class NodeData(Serializable):
    
    # "None", "Start", "Agent", "Task", "Step", "Team"
    type: str = ""

    uniq_id: str = ""
    pos_x: float = 0.0
    pos_y: float = 0.0
    width: float = 200.0  # Default width
    height: float = 200.0  # Default height

    name: str = ""

    # Agent
    role: str = ""
    goal: str = ""
    backstory: str = ""

    # Task
    agent: str = ""
    description: str = ""
    expected_output: str = ""

    # Step
    tool: str = ""
    arg: str = ""
    output_var: str = ""


    nexts: List[int] = field(default_factory=list)
    prevs: List[int] = field(default_factory=list)

    def to_dict(self):
        return asdict(self)

    @classmethod
    def from_dict(cls, data):
        return cls(**data)

File: crewai-reactflow/src/components/NodeData.js
// nodeData.js
class NodeData {
    constructor(
      uniq_id,
      type = "",
      pos_x = 0.0,
      pos_y = 0.0,
      width = 200.0,
      height = 200.0,
      name = "",
      role = "",
      goal = "",
      backstory = "",
      agent = "",
      description = "",
      expected_output = "",
      tool = "",
      arg = "",
      output_var = "",
      nexts = [],
      prevs = []
    ) {
      this.uniq_id = uniq_id;
      this.type = type;
      this.pos_x = pos_x;
      this.pos_y = pos_y;
      this.width = width;
      this.height = height;
      this.name = name;
      this.role = role;
      this.goal = goal;
      this.backstory = backstory;
      this.agent = agent;
      this.description = description;
      this.expected_output = expected_output;
      this.tool = tool;
      this.arg = arg;
      this.output_var = output_var;
      this.nexts = nexts;
      this.prevs = prevs;
    }
  
    toDict() {
      return { ...this };
    }
  
    static fromDict(data) {
      return new NodeData(
        data.uniq_id,
        data.type,
        data.pos_x,
        data.pos_y,
        data.width,
        data.height,
        data.name,
        data.role,
        data.goal,
        data.backstory,
        data.agent,
        data.description,
        data.expected_output,
        data.tool,
        data.arg,
        data.output_var,
        data.nexts,
        data.prevs
      );
    }
  }
  
  export default NodeData;
  